File: app/src/main/java/com/adsamcik/signalcollector/Preferences.java
package com.adsamcik.signalcollector;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.adsamcik.signalcollector.data.StatDay;
import com.adsamcik.signalcollector.services.TrackerService;
import com.google.gson.Gson;

import java.util.HashSet;
import java.util.Set;

public class Preferences {
	private static final String TAG = "SignalsSetting";
	public static final int UPLOAD_JOB = 513;

	public static final String SCHEDULED_UPLOAD = "uploadSCHEDULED";
	public static final String SENT_TOKEN_TO_SERVER = "sentTokenToServer";
	public static final String REGISTERED_USER = "playUserRegistered";
	public static final String BACKGROUND_TRACKING = "backgroundTracking";
	public static final String AUTO_UPLOAD = "autoUpload";
	public static final String HAS_BEEN_LAUNCHED = "hasBeenLaunched";
	public static final String STOP_TILL_RECHARGE = "stoppedTillRecharge";

	//Local tracking stats
	public static final String STATS_WIFI_FOUND = "statsWifiFound";
	public static final String STATS_CELL_FOUND = "statsCellFound";
	public static final String STATS_LOCATIONS_FOUND = "statsLocationsFound";
	public static final String STATS_MINUTES = "statsMinutes";
	public static final String STATS_STAT_LAST_DAY = "statsLastDay";
	public static final String STATS_LAST_7_DAYS = "statsLast7Days";

	public static final String STATS_UPLOADED = "statsUploaded";
	//obsolete
	public static final String STATS_UPLOADED_OLD = "statsUploadKilobytes";

	public static final String AVAILABLE_MAPS = "availableMaps";

	public static final String OLDEST_RECENT_UPLOAD = "oldestRecentUpload";

	public static final String GENERAL_STATS_LAST_UPDATE = "generalStatsLastUpdate";
	public static final String USER_STATS_LAST_UPDATE = "userStatsLastUpdate";
	public static final String AVAILABLE_MAPS_LAST_UPDATE = "availableMapsLastUpdate";

	public static final String TRACKING_WIFI_ENABLED = "trackingWifiEnabled";
	public static final String TRACKING_CELL_ENABLED = "trackingCellEnabled";
	public static final String TRACKING_NOISE_ENABLED = "trackingNoiseEnabled";

	public static final String NOTIFICATIONS_ENABLED = "notificationsEnabled";

	private static final int MAX_DAY_DIFF_STATS = 7;

	private static SharedPreferences sharedPreferences;

	/**
	 * Will stop tracking until phone is connected to charger
	 *
	 * @param c context
	 */
	public static void stopTillRecharge(@NonNull Context c) {
		get(c).edit().putBoolean(STOP_TILL_RECHARGE, true).apply();
		if (TrackerService.service != null)
			c.stopService(TrackerService.service);
	}

	/**
	 * Get shared preferences
	 * This function should never crash. Initializes preferences if needed.
	 *
	 * @param c Non-null context
	 * @return Shared preferences
	 */
	public static SharedPreferences get(@NonNull Context c) {
		if (sharedPreferences == null)
			sharedPreferences = PreferenceManager.getDefaultSharedPreferences(c.getApplicationContext());
		return sharedPreferences;
	}

	/**
	 * Get shared preferences
	 * This function will crash if shared preferences were never initialized
	 * Always prefer to send context if posssible.
	 *
	 * @return Shared preferences
	 */
	public static SharedPreferences get() {
		if (sharedPreferences == null)
			throw new RuntimeException("Shared preferences are null and no context was provided");
		return sharedPreferences;
	}

	public static void checkStatsDay(@Nullable Context context) {
		if(sharedPreferences == null) {
			if(context == null)
				throw new RuntimeException("Shared preferences and context are null");
			else
				sharedPreferences = get(context);
		}
		long todayUTC = Assist.getDayInUTC();
		int dayDiff = (int) (todayUTC - sharedPreferences.getLong(Preferences.STATS_STAT_LAST_DAY, -1)) / Assist.DAY_IN_MILLISECONDS;
		if (dayDiff > 0) {
			Set<String> stringStats = sharedPreferences.getStringSet(STATS_LAST_7_DAYS, null);
			Set<StatDay> stats = fromJson(stringStats, dayDiff);

			stats.add(getCurrent(sharedPreferences));

			if (stringStats == null)
				stringStats = new HashSet<>();
			else
				stringStats.clear();

			Gson gson = new Gson();
			for (StatDay day : stats)
				stringStats.add(gson.toJson(day));

			sharedPreferences.edit()
					.putLong(STATS_STAT_LAST_DAY, todayUTC)
					.putStringSet(STATS_LAST_7_DAYS, stringStats)
					.putInt(STATS_MINUTES, 0)
					.putInt(STATS_LOCATIONS_FOUND, 0)
					.putInt(STATS_WIFI_FOUND, 0)
					.putInt(STATS_CELL_FOUND, 0)
					.putInt(STATS_CELL_FOUND, 0)
					.apply();
		}
	}

	public static StatDay countStats(@NonNull Context context) {
		SharedPreferences sp = get(context);
		StatDay result = getCurrent(sp);
		Set<StatDay> set = fromJson(sp.getStringSet(STATS_LAST_7_DAYS, null), 0);
		//noinspection Convert2streamapi
		for (StatDay stat : set)
			result.add(stat);
		return result;
	}

	private static StatDay getCurrent(SharedPreferences sp) {
		return new StatDay(sp.getInt(STATS_MINUTES, 0), sp.getInt(STATS_LOCATIONS_FOUND, 0), sp.getInt(STATS_WIFI_FOUND, 0), sp.getInt(STATS_CELL_FOUND, 0), 0, sp.getLong(STATS_UPLOADED, 0));
	}

	private static Set<StatDay> fromJson(Set<String> set, int age) {
		Set<StatDay> statDays = new HashSet<>();
		if (set == null)
			return statDays;
		Gson gson = new Gson();
		for (String val : set) {
			StatDay sd = gson.fromJson(val, StatDay.class);
			if (age <= 0 || sd.age(age) < MAX_DAY_DIFF_STATS)
				statDays.add(sd);
		}
		return statDays;
	}

}

File: utilities/src/main/java/com/adsamcik/utilities/Preferences.kt
package com.adsamcik.utilities

import android.app.Activity
import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.adsamcik.signalcollector.data.StatDay
import com.adsamcik.signalcollector.utility.Constants.DAY_IN_MILLISECONDS
import com.google.gson.Gson
import java.util.*

object Preferences {
    private const val TAG = "SignalsSetting"

    const val PREF_THEME = "theme"
    const val DEFAULT_THEME = R.style.AppThemeLight

    const val LAST_VERSION = "lastVersion"

    const val PREF_SCHEDULED_UPLOAD = "uploadSCHEDULED"
    const val PREF_SENT_TOKEN_TO_SERVER = "sentTokenToServer"
    const val PREF_REGISTERED_USER = "playUserRegistered"

    const val PREF_AUTO_TRACKING = "backgroundTracking"
    const val DEFAULT_AUTO_TRACKING = 1
    const val PREF_AUTO_UPLOAD = "autoUpload"
    const val DEFAULT_AUTO_UPLOAD = 1

    const val PREF_HAS_BEEN_LAUNCHED = "hasBeenLaunched"
    const val PREF_STOP_TILL_RECHARGE = "stoppedTillRecharge"

    const val PREF_USER_ID = "userID"

    const val PREF_AUTO_UPLOAD_AT_MB = "autoUpAtMB"
    const val DEFAULT_AUTO_UPLOAD_AT_MB = 3

    const val PREF_AUTO_UPLOAD_SMART = "smartAutoUpload"
    const val DEFAULT_AUTO_UPLOAD_SMART = true

    //Local tracking stats
    const val PREF_STATS_WIFI_FOUND = "statsWifiFound"
    const val PREF_STATS_CELL_FOUND = "statsCellFound"
    const val PREF_STATS_LOCATIONS_FOUND = "statsLocationsFound"
    const val PREF_STATS_MINUTES = "statsMinutes"
    const val PREF_STATS_STAT_LAST_DAY = "statsLastDay"
    const val PREF_STATS_LAST_7_DAYS = "statsLast7Days"
    const val PREF_STATS_UPLOADED = "statsUploaded"

    const val PREF_OLDEST_RECENT_UPLOAD = "oldestRecentUpload"

    const val PREF_GENERAL_STATS = "generalStats"
    const val PREF_STATS = "stats"
    const val PREF_USER_STATS = "userStats"
    const val PREF_AVAILABLE_MAPS = "availableMaps"

    const val PREF_USER_DATA = "userDATA"
    const val PREF_USER_PRICES = "userPRICES"

    const val PREF_TRACKING_WIFI_ENABLED = "trackingWifiEnabled"
    const val DEFAULT_TRACKING_WIFI_ENABLED = true
    const val PREF_TRACKING_CELL_ENABLED = "trackingCellEnabled"
    const val DEFAULT_TRACKING_CELL_ENABLED = true
    const val PREF_TRACKING_NOISE_ENABLED = "trackingNoiseEnabled"
    const val DEFAULT_TRACKING_NOISE_ENABLED = false
    const val PREF_TRACKING_LOCATION_ENABLED = "trackingLocationEnabled"
    const val DEFAULT_TRACKING_LOCATION_ENABLED = true

    const val PREF_ACTIVITY_WATCHER_ENABLED = "activityWatcherEnabled"
    const val DEFAULT_ACTIVITY_WATCHER_ENABLED = false
    const val PREF_ACTIVITY_UPDATE_RATE = "activityUpdateRate"
    const val DEFAULT_ACTIVITY_UPDATE_RATE = 30

    const val PREF_COLLECTIONS_SINCE_LAST_UPLOAD = "collectionsSinceLastUpload"
    const val PREF_COLLECTIONS_IN_STORAGE = "collectionsInStorage"

    const val PREF_DEFAULT_MAP_OVERLAY = "defaultMapOverlay"

    const val PREF_UPLOAD_NOTIFICATIONS_ENABLED = "uploadNotificationsEnabled"

    const val PREF_SHOW_DEV_SETTINGS = "showDevSettings"
    const val PREF_DEV_ACTIVITY_TRACKING_ENABLED = "activityDebugTracking"
    const val PREF_DEV_ACTIVITY_TRACKING_STARTED = "adevTrackingStartTime"

    const val PREF_ACTIVE_CHALLENGE_LIST = "activeChallengeList"

    private const val MAX_DAY_DIFF_STATS = 7

    private var sharedPreferences: SharedPreferences? = null

    /**
     * Get shared preferences
     * This function should never crash. Initializes preferences if needed.
     *
     * @param c Non-null context
     * @return Shared preferences
     */
    fun getPref(c: Context): SharedPreferences {
        if (sharedPreferences == null)
            sharedPreferences = PreferenceManager.getDefaultSharedPreferences(c.applicationContext)
        return sharedPreferences!!
    }

    fun getTheme(context: Context): Int = getPref(context).getInt(PREF_THEME, DEFAULT_THEME)

    fun setTheme(activity: Activity, theme: Int) {
        getPref(activity).edit().putInt(PREF_THEME, theme).apply()
        activity.setTheme(theme)
        //This ensures that all components use the proper theme
        activity.applicationContext.setTheme(theme)
    }

    fun setTheme(activity: Activity) {
        val theme = getTheme(activity)
        activity.setTheme(theme)
        //This ensures that all components use the proper theme
        activity.applicationContext.setTheme(theme)
    }

    /**
     * Checks if current day should be archived and clears up old StatDays
     * @param context context
     */
    fun checkStatsDay(context: Context) {
        val preferences = getPref(context)

        val todayUTC = Assist.todayUTC
        val dayDiff = (todayUTC - preferences.getLong(PREF_STATS_STAT_LAST_DAY, -1)).toInt() / DAY_IN_MILLISECONDS
        if (dayDiff > 0) {
            var stringStats: MutableSet<String>? = preferences.getStringSet(PREF_STATS_LAST_7_DAYS, null)
            val stats = fromJson(stringStats, dayDiff)

            stats.add(getCurrent(preferences))

            if (stringStats == null)
                stringStats = HashSet()
            else
                stringStats.clear()

            val gson = Gson()
            stats.mapTo(stringStats) { gson.toJson(it) }

            preferences.edit()
                    .putLong(PREF_STATS_STAT_LAST_DAY, todayUTC)
                    .putStringSet(PREF_STATS_LAST_7_DAYS, stringStats)
                    .putInt(PREF_STATS_MINUTES, 0)
                    .putInt(PREF_STATS_LOCATIONS_FOUND, 0)
                    .putInt(PREF_STATS_WIFI_FOUND, 0)
                    .putInt(PREF_STATS_CELL_FOUND, 0)
                    .putLong(PREF_STATS_UPLOADED, 0)
                    .apply()
        }
    }

    /**
     * Counts all stats and combines them to a single StatDay object
     * @param context context
     * @return sum of all StatDays
     */
    fun countStats(context: Context): StatDay {
        val sp = getPref(context)
        val result = getCurrent(sp)
        val set = fromJson(sp.getStringSet(PREF_STATS_LAST_7_DAYS, null), 0)

        result += set
        return result
    }

    /**
     * Creates stat day with today values
     * @param sp shared preferences
     * @return Today StatDay
     */
    private fun getCurrent(sp: SharedPreferences): StatDay =
            StatDay(sp.getInt(PREF_STATS_MINUTES, 0), sp.getInt(PREF_STATS_LOCATIONS_FOUND, 0), sp.getInt(PREF_STATS_WIFI_FOUND, 0), sp.getInt(PREF_STATS_CELL_FOUND, 0), 0, sp.getLong(PREF_STATS_UPLOADED, 0))

    /**
     * Method loads data to list and checks if they are not too old
     * @param set string set
     * @param age how much should stats age
     * @return list with items that are not too old
     */
    private fun fromJson(set: Set<String>?, age: Int): MutableList<StatDay> {
        val statDays = ArrayList<StatDay>()
        if (set == null)
            return statDays
        val gson = Gson()
        set
                .map { gson.fromJson(it, StatDay::class.java) }
                .filterTo(statDays) { age <= 0 || it.age(age) < MAX_DAY_DIFF_STATS }
        return statDays
    }
}
